Speaker Feedback: Register API endpoints for feedback comments

coreymckrill opened the pull request:
Extends the `WP_REST_Comments_Controller` class to create a new custom endpoint for creating feedback comments.

We can't use the standard Comments endpoint, because it won't allow creation of comments with a custom type or without anything in the content field.

This also introduces schema and functions for the comment meta values that we'll be using to store feedback content, as well as sets up this plugin for phpunit testing.

Fixes #337, Refs #340 

**To test:**

1. Run all the phpunit tests and make sure all are passing.
1. It's easiest to use a tool like [Insomnia](https://insomnia.rest/) to craft payloads to send to the endpoint. The endpoint URL will be `[your-test-site].wordcamp.org/wp-json/wordcamp-speaker-feedback/v1/feedback`
1. The parameters required for a successful payload are:
    * `post` (the ID of the post to attach the feedback comment to)
    * `author` (the numeric ID of a user)
    * `meta[rating]` (an integer representation of 1 to 5 stars)
    * `meta[q1]` (a string less than 5000 characters)
1. You can use `author_name` and `author_email` (string values) in lieu of `author`.
1. You can also include `meta[q2]` and `meta[q3]` parameters, but these are not currently required fields.
1. The response to a successful request to the endpoint should be a `201` status. For an unsuccessful request, it'll be a JSON object with an error code and message. Try to get some of each! Try multiple invalid fields in one request!
1. It would probably also be good to check that the comments and their meta are actually being created in the database upon successful request.

coreymckrill committed:
Add REST controller scaffold and initialize routes

coreymckrill committed:
REST Controller: limit to creating new feedback submissions

At this point, we don't need the full complement of endpoints for
feedback, we just need a way to create new feedback comments from
the front end.

coreymckrill committed:
Add functions for managing comment meta

coreymckrill committed:
Code cleanup

coreymckrill committed:
Add meta version before validating

coreymckrill committed:
Change comment type string to fit database limit

coreymckrill committed:
Add helper functions for feedback comments

coreymckrill committed:
Use new get_feedback_comment helper in get_feedback function

coreymckrill committed:
Update doc blocks in comment.php

coreymckrill committed:
Tweak string length testing in meta validation

coreymckrill committed:
Set up php unit testing

coreymckrill committed:
Add unit tests for comment.php

coreymckrill committed:
Add unit tests for comment-meta.php

coreymckrill committed:
Meta: Add validation for integer fields, related unit test

coreymckrill committed:
Clean up create_item()

coreymckrill committed:
Add initial unit test for REST Feedback Controller

coreymckrill committed:
Tests: Add more for rest controller, improve teardown

coreymckrill committed:
Tests: improve teardown for comments

coreymckrill commented:
Bah, the tests are passing locally...

ryelle commented:
> Bah, the tests are passing locally...

for what it's worth, the test failed for me the same was as travis the first time I ran them, but have worked each time since.

coreymckrill committed:
Tests: Refactor to use `wpSetUpBeforeClass` instead of `__construct`

Creating fixtures (posts, users) during class instantiation
(__construct) was causing a weird issue where by the time a particular
test ran that required a user ID, that user no longer existed. Whereas,
`wpSetUpBeforeClass` is called just before the first test in the class
is run.

coreymckrill committed:
Normalize error codes in validate_feedback_meta

coreymckrill committed:
Update comment meta tests

coreymckrill committed:
Update comment tests

coreymckrill committed:
REST controller: Clean up error codes, add more comments

coreymckrill committed:
Update rest controller tests

ryelle commented:
I'm having trouble testing this with Insomnia, it keeps giving me this response even though I know the ID I've given it is a real session.

```json
{
  "code": "rest_feedback_invalid_post_id",
  "message": "Sorry, you are not allowed to create this comment without a post.",
  "data": {
    "status": 403
  }
}
```

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/class-rest-feedback-controller.php:
```diff
@@ -0,0 +1,225 @@
+		if ( empty( $request['post'] ) ) {
+			return new WP_Error(
+				'rest_feedback_invalid_post_id',
+				__( 'Sorry, you are not allowed to create this comment without a post.' ),
```
This is missing a textdomain

coreymckrill replied:
Oops, good catch. Fixed here: 025f85f8

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/class-rest-feedback-controller.php:
```diff
@@ -0,0 +1,225 @@
+		if ( ! $post ) {
+			return new WP_Error(
+				'rest_feedback_invalid_post_id',
+				__( 'Sorry, you are not allowed to create this comment without a post.' ),
```
This message confused me at first since I did give it a post — maybe something about needing a valid post. Should it also check for Session post type? (also, missing textdomain here too)

coreymckrill replied:
I'm not sure why it's not recognizing your post ID. I just tried it again with Insomnia and it's still working for me...

> Should it also check for Session post type?

Yeah, I was thinking of including that as part of another validation function that would go where the `TODO` note currently is. I've been thinking I would add that as part of a separate PR for #340, but I guess that one is specifically about permissions for viewing/getting, so perhaps it would be better to do it here. Stay tuned.

coreymckrill replied:
Added post type support and a check for it here: 2c9036af

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment-meta.php:
```diff
@@ -0,0 +1,207 @@
+ */
+function get_feedback_meta_field_schema( $key = '' ) {
+	$schema = array(
+		'version' => array(
```
Good idea with versioning the questions! 💯

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment-meta.php:
```diff
@@ -0,0 +1,207 @@
+	$integer_fields = array_keys( wp_list_pluck( $fields, 'type' ), 'integer', true );
+
+	foreach ( $integer_fields as $key ) {
+		if ( isset( $meta[ $key ] ) && ! is_numeric( $meta[ $key ] ) ) {
```
Can we set some custom values in the schema? Thinking it might be good to check against a "max" so that no one tries to enter in a rating of 100 or something (in a future PR is fine)

coreymckrill replied:
See what you think of this: 42180384

ryelle reviewed (commented):
some notes from looking through the code — overall, great improvements here; dunno why it's not working for me. It seems like the endpoint `$request` is not actually receiving the parameters. I'll try testing again when I'm back on Friday.

coreymckrill committed:
Add missing translation and textdomains

coreymckrill committed:
Add post type support

coreymckrill committed:
Add support check to create_item_permissions_check() and test

coreymckrill committed:
Meta: Add attributes to field schema, refactor validation

Attributes can be used for both validation and for related HTML input
elements.

Validation now takes field type into account and looks for a
corresponding validation callback. If it determines that the submitted
data for the field is invalid, it returns a WP_Error object containing
the field key and an error message. These are then compiled and returned
as a single WP_Error object with all invalid data keys and messages
when validation is complete. This way we can show specific error
messages on specific fields on the form.

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/wordcamp-speaker-feedback.php:
```diff
@@ -50,6 +60,20 @@ function activate() {
+ * Add post type support for supported post types. Only for the types that are supported though.
+ *
+ * This makes it easy to check if a particular post type can have feedback comments using Core functionality, rather
+ * than having to import a namespaced constant or function.
```
Also a great idea 😁

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/comment-meta.php:
```diff
@@ -0,0 +1,295 @@
+				'step' => 1,
+			),
+		),
+		'q1'      => array(
```
The first question is also required:

> The talk selection, star rating and first text area will be required fields with the rest being optional. [(source)](https://make.wordpress.org/community/2020/01/27/speaker-feedback-tool-final-outline/)

coreymckrill replied:
Fixed: f92715ae

ryelle commented on public_html/wp-content/plugins/wordcamp-speaker-feedback/includes/class-rest-feedback-controller.php:
```diff
@@ -0,0 +1,235 @@
+			);
+		}
+
+		$response = __( 'Success!', 'wordcamporg' );
```
Will a user ever see this? I'd expect this endpoint to only be called by javascript, so we'd want a stable return value to check against — a string that might be in a different language could make that tricky.

coreymckrill replied:
Good point. It does also send a `201` response code, which should be stable. But `Success!` probably isn't necessary.

The default comment endpoint returns data about the comment on successful creation. I wasn't sure we would want to do that in this case since feedback comments are not public. What do you think would be best?

ryelle replied:
yeah, I'm agreed on not returning comment data here, I don't think that's necessary. I didn't think about the response code, so that should be enough. Does the response need content? we might be fine with an empty response (unless that does unexpected things).

coreymckrill replied:
I'll test and find out...

coreymckrill replied:
Seems to work fine. I updated the response here: 3415a674

ryelle reviewed (approved):
Turns out my API issues were user-error 🤦‍♀  All working as expected now. I really like the validation approach 💯 

Two minor comments, then this is ready to merge 👍

coreymckrill committed:
Remove unneeded META_MAX_LENGTH constant

coreymckrill committed:
Make q1 required

coreymckrill committed:
Tweak the REST response for successful feedback creation

coreymckrill committed:
Fix tests
